src/python/torchdistx/slow_momentum/slow_momentum_optimizer.py

```diff
@@ -0,0 +1,222 @@
+        >>>  # set `sync_grads` to `False`
+        >>>  cur_subgroup, _ = dist.new_subgroups()
+        >>>  slowMoState = slow_momentum_comm.SlowMoState(
+        >>>     cur_subgroup,
+        >>>     sync_grads=True
+        >>>  )
+        >>>
+        >>>  # Register SlowMo hook, which only communicates gradients
+        >>>  # in a intra-node fashion.
+        >>>  fsdp_net.register_comm_hook(
+        >>>     slowMoState,
+        >>>     slow_momentum_comm.slowmo_hook
+        >>>  )
+        >>>
+        >>>  base_optimizer = torch.optim.SGD(
+        >>>     fsdp_net_slowmo.parameters(),
+        >>>     lr=1e-2
+        >>>  )
+        >>>  # Create a SlowMo optimizer that wraps a local optimizer.
+        >>>  slowmo_optim = slow_momentum_optimizer.SlowMomentumOptimizer(
+        >>>     base_optim=base_optimizer,
+        >>>     slowmo_freq=6,
+        >>>     slowmo_factor=0.5,
+        >>>     slowmo_lr=0.1
+        >>>  )
+        >>>
+        >>>  # SlowMo runs intra-node gradient averaging at every step,
+        >>>  # every 6th step it will run model averaging and
+        >>>  # a slow momentum update.
+        >>>  for step in range(0, 200):
+        >>>     slowmo_optim.zero_grad()
+        >>>     loss = loss_fn(output, labels)
+        >>>     loss.backward()
+        >>>     slowmo_optim.step()
+    """
+
+    def __init__(
+        self,
+        base_optim: torch.optim.Optimizer,
+        slowmo_freq: int = 48,
+        slowmo_factor: float = 0.5,
+        slowmo_lr: float = 1.0,
+    ):
+        if base_optim is None:
+            raise ValueError("Base optimizer is a required parameter.")
+        self._base_optim = base_optim
+
+        # check that base optimizer's learning rate is stored in param_groups
+        if not (
+            self._base_optim.param_groups and self._base_optim.param_groups[0]["lr"]
+        ):
+            raise ValueError(
+                "Provided base optimizer does not have "
+                "parameters or learning rate specified."
+            )
+        self.param_groups = self._base_optim.param_groups
+        self.base_lr = self.param_groups[0]["lr"]
+
+        if slowmo_freq < 1:
+            raise ValueError(
+                "Invalid ``slowmo_freq`` parameter, must be a positive value."
+            )
+        self.slowmo_freq = slowmo_freq
+
+        if slowmo_factor < 0.0:
+            raise ValueError(
+                "Invalid ``slowmo_factor`` parameter, must be non-negative."
+            )
+        self.slowmo_factor = slowmo_factor
+
+        if slowmo_lr < 0.0:
+            raise ValueError("Invalid ``slowmo_lr`` parameter, must be non-negative.")
+        self.slowmo_lr = slowmo_lr
+
+        self.averager = averagers.PeriodicModelAverager(
+            period=slowmo_freq, warmup_steps=0
+        )
+        self._init_slowmo_buffers()
+
+    def _init_slowmo_buffers(self):
+        for group in self.param_groups:
```

Comment:
Do we need to run `self.state.clear()` at the beginning of this?

(This is relevant when this method is called in `load_state_dict()`. It seems like we may want a clean slate.)

Reply:
@awgu  We actually don't want to clear `state`, because for example Adam will have specific parameters stored in there, and if we clear `state`, we clear that parameters